### PR TITLE
Additional Vault Authentication method

### LINKED ISSUE #61243: [BUG] salt.utils.aws.sig4 force usage of version as GET parameters even if method is not GET
**Description**

** Setup
No Setup required

**Steps to Reproduce the behavior**
```
from salt.utils.aws import sig4
import requests

method = "POST"
body="Action=GetCallerIdentity&Version=2011-06-15"
endpoint="sts.amazonaws.com"
location="us-east-1"
params = {}
version = "2011-06-15"
product = "sts"
provider = {'id': 'use-instance-role-credentials', 'key': 'use-instance-role-credentials'}
hd = {"Content-Type": "application/x-www-form-urlencoded; charset=utf-8"}
headers, requests_url = sig4(   method=method, 
                                endpoint=endpoint,
                                prov_dict=provider,
                                params=params,
                                product="sts",
                                data=body,
                                aws_api_version=version,
                                location=location,
                                requesturl=f"https://{endpoint}/",
                                headers=hd
                            )
print(requests_url)
# https://sts.amazonaws.com?Version=2011-06-15
print(headers)
# {'Content-Type': 'application/x-www-form-urlencoded; charset=utf-8', 'X-Amz-date': '20211116T160158Z', 'host': 'sts.amazonaws.com', 'x-amz-content-sha256': 'ab821ae955788b0e33ebd34c208442ccfc2d406e2edc5e7a39bd6458fbb4f843', 'X-Amz-security-token': '<redacted>', 'Authorization': 'AWS4-HMAC-SHA256 Credential=<redacted>/20211116/us-east-1/sts/aws4_request, SignedHeaders=content-type;host;x-amz-content-sha256;x-amz-date;x-amz-security-token, Signature=<redacted>'}
r = requests.post(requests_url,headers= headers, data=body)
print(r.status_code)
# 400
print(r.text)
# ... When Content-Type:application/x-www-form-urlencoded, URL cannot include query-string parameters (after \'?\'): \'/?Version=2011-06-15\' ...

params = {"Action":"GetCallerIdentity", "Version": "2011-06-15"}
method = "GET"
hd = {}
body=""
headers, requests_url = sig4(   method=method, 
                                endpoint=endpoint,
                                prov_dict=provider,
                                params=params,
                                product="sts",
                                data=body,
                                aws_api_version=version,
                                location=location,
                                requesturl=f"https://{endpoint}/",
                                headers=hd
                            )
print(requests_url)
# https://sts.amazonaws.com/?Action=GetCallerIdentity&Version=2011-06-15
r = requests.get(requests_url,headers= headers)
print(r.status_code)
# 200
```

**Expected behavior**
request_url should not contains any params when not using GET method and then query_string should be empty in order to have a valid aws sig4 signature

**Versions Report**
<details><summary>salt --versions-report</summary>
```
Salt Version:
          Salt: 3004

Dependency Versions:
          cffi: Not Installed
      cherrypy: Not Installed
      dateutil: 2.7.3
     docker-py: Not Installed
         gitdb: 2.0.6
     gitpython: 3.0.7
        Jinja2: 2.10.1
       libgit2: 0.28.3
      M2Crypto: Not Installed
          Mako: Not Installed
       msgpack: 0.6.2
  msgpack-pure: Not Installed
  mysql-python: Not Installed
     pycparser: Not Installed
      pycrypto: Not Installed
  pycryptodome: 3.6.1
        pygit2: 1.0.3
        Python: 3.8.10 (default, Sep 28 2021, 16:10:42)
  python-gnupg: 0.4.5
        PyYAML: 5.3.1
         PyZMQ: 18.1.1
         smmap: 2.0.5
       timelib: Not Installed
       Tornado: 4.5.3
           ZMQ: 4.3.2

System Versions:
          dist: ubuntu 20.04 focal
        locale: utf-8
       machine: x86_64
       release: 5.11.0-1020-aws
        system: Linux
       version: Ubuntu 20.04 focal
```
</details>

### PR DESCRIPTION
**Pull request by johnvox**
### What does this PR do?

This PR add the following authentication method for Hashicorp Vault modules:
- AWS
- Kubernetes

At the same time in order to used already wrote code a small fix has been included in AWS utils in order to be able to sign other type of requests (currently only GET worked fine)

### What issues does this PR fix or reference?
Fixes: #61243 

### New Behavior
It's now possible to leverage AWS or Kubernetes for the integration between salt and vault. It not need to used addition credentials when running in one of those environment

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**

- [x] Docs
- [X] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [X] Tests written/updated

### Commits signed with GPG?
Yes

### CODE REVIEW

File: tests/unit/runners/test_vault.py
Comment (waynew) [inline tests/unit/runners/test_vault.py:444]: **suggestion**Patches already create mocks. Also it's **usually** preferable to use `autospec=True`:

```
@patch("salt.runners.vault._validate_signature", autospec=True, return_value=None)
```

This suggestion is broadly applicable.

**Review by waynew — CHANGES_REQUESTED** [review]

File: tests/unit/runners/test_vault.py
Comment (jynolen) [reply tests/unit/runners/test_vault.py:444]: Hello can you please explain me a bit more not sure to understand it completly.
To write new tests I just take the same approach of upstream. 
[https://github.com/saltstack/salt/blob/094687af0565c2e410b77e3d35de3c9807dd404c/tests/unit/runners/test_vault.py#L338](url) doesn't use `autospec=True` neither.

Should we change upstream testcase as well ? :/

**Comment by jynolen** [comment]
Any updates for this ?

File: tests/unit/runners/test_vault.py
Comment (jynolen) [reply tests/unit/runners/test_vault.py:444]: Hello.
Can you have a look to my previous comment ? thank you

Comment (waynew) [reply tests/unit/runners/test_vault.py:444]: Hey @jynolen sorry for the extremely late reply, apparently this PR got lost in my GitHub noise :disappointed: 

`autospec=True` is almost always a good thing, because it will match the function signatures. For example:

```
def add(a, b):
    ...
    
with patch("__main__.add", return_value=42):
    fake_add(3, 4, 10, 100)
```

That will work, because MagicMock's are extraordinarily flexible. But with `autospec=True` you'd get a `TypeError: too many positional arguments`.

There are some exceptions and reasons you might not want to autospec, but if you're making a mock then almost always autospec is the right way to go.

I would say that it's not *necessary* to change the other tests, but I would definitely approve of changing the other tests (assuming that they're not one of the exceptions :joy:)

**Comment by waynew** [comment]
@jynolen looks like there's a merge conflict in the test.

**Comment by waynew** [comment]
Hey @jynolen since there's not been any activity on this PR I'm going to go ahead and close it for now - if you (or anyone else) want to pick up this PR  and get it back across the finish line just let myself or another core team member know - you can typically find us on IRC or Slack - and we can re-open this PR :+1:

**Pull request closed without merging**
